### PR TITLE
Reduce temporary allocations

### PR DESCRIPTION
**Pull request by wxmerkt**
- CostModelResidual: Workaround for heap allocation in the product with the diagonal weighting matrix of the activation. This is a workaround for https://gitlab.com/libeigen/eigen/-/issues/2408
- LQR: Introduces a temporary variable for the Hessian-vector product to eliminate temporaries in the computation of the cost

### CODE REVIEW

File: include/crocoddyl/core/actions/lqr.hxx
Comment (cmastalli) [inline include/crocoddyl/core/actions/lqr.hxx:81]: Same here as comment below

File: include/crocoddyl/core/actions/lqr.hpp
Comment (cmastalli) [inline include/crocoddyl/core/actions/lqr.hpp:96]: Please construct the size of the matrix and use `setZero`, e.g,
```c++
  explicit ActionDataLQRTpl(Model<Scalar>* const model) : Base(model), Luu_u(model->get_nu()) {
  Luu_u.setZero();
}
```

Note also that I have renamed the variable to `Luu_u`. Do the same for `Hx_tmp` (i.e., `Lxu_u`).

Further comment:
This notation is not allowed in Crocoddyl 2 (to be release asap). Those names are reserved to store constraint derivatives.

File: include/crocoddyl/core/actions/lqr.hxx
Comment (cmastalli) [inline include/crocoddyl/core/actions/lqr.hxx:41]: We don't accept this code format. Similar to other comments.

File: include/crocoddyl/core/actions/lqr.hxx
Comment (cmastalli) [inline include/crocoddyl/core/actions/lqr.hxx:71]:
```suggestion
d->Lxx_x.noalias() = Lxx_ * x;
d->Luu_x.noalias() = Luu_ * u;
data->cost = Scalar(0.5) * x.dot(d->Lx_x);
data->cost += Scalar(0.5) * x.dot(d->Lx_u);
d->Lxx_x.noalias() = Lxu_ * u;
data->cost += Scalar(0.5) * x.dot(d->Lx_x);
data->cost += lx_.dot(x);
data->cost += lu_.dot(u);
```

Additionally, I think we could add some `dot` in the same line. I think this should not allocate memory as the results is a scalar. Am I right?

File: include/crocoddyl/core/costs/residual.hxx
Comment (cmastalli) [inline include/crocoddyl/core/costs/residual.hxx:63]: Could you provide information about improve performance?
I ask this as the results reported in private where not better (in terms of computation time).

**Review by cmastalli — CHANGES_REQUESTED** [review]

File: include/crocoddyl/core/actions/lqr.hxx
Comment (wxmerkt) [reply include/crocoddyl/core/actions/lqr.hxx:104]: I did run it (and discussed with Guilhem) -- this is what the formatter caused :-(

File: include/crocoddyl/core/actions/lqr.hpp
Comment (wxmerkt) [reply include/crocoddyl/core/actions/lqr.hpp:96]: Okay, will do - in this case we will have to add an additional variable resulting in a slightly larger struct (Lxx_x, Lxu_u, Luu_u)

File: include/crocoddyl/core/actions/lqr.hxx
Comment (wxmerkt) [reply include/crocoddyl/core/actions/lqr.hxx:104]: I.e. the formatter does not consider this to be incorrect / does not fix it

File: include/crocoddyl/core/costs/residual.hxx
Comment (cmastalli) [reply include/crocoddyl/core/costs/residual.hxx:63]: Ok, thanks for the info.

File: include/crocoddyl/core/actions/lqr.hpp
Comment (cmastalli) [reply include/crocoddyl/core/actions/lqr.hpp:96]: Lxx_x` and `Lxu_u` have the same dimension, you could reuse it.

Actually, it might be better to rename them as `Lxx_x_tmp` and `Luu_u_tmp`, as `Lxx_x` is not necessary `Lxx * x`.

File: include/crocoddyl/core/actions/lqr.hxx
Comment (cmastalli) [reply include/crocoddyl/core/actions/lqr.hxx:104]: @wxmerkt 
Did you use `clang-6`?
This code is not c++98 compatible; thus, it is weird that this is the case.

@nim65s 
Why is now different than before? 
Are we using other setup in the CI?

Comment (wxmerkt) [reply include/crocoddyl/core/actions/lqr.hxx:104]: From what I recall we moved to C++14 already since we implicitly had that requirement anyways (https://github.com/loco-3d/crocoddyl/issues/934), so it shouldn't matter if it's not C++98 compatible:

https://github.com/loco-3d/crocoddyl/blob/9e79f80c6d52b2c6c000106570811d449341cad3/CMakeLists.txt#L63

File: include/crocoddyl/core/actions/lqr.hxx
Comment (wxmerkt) [reply include/crocoddyl/core/actions/lqr.hxx]: Removed comments and amended with dot

File: include/crocoddyl/core/actions/lqr.hxx
Comment (cmastalli) [inline include/crocoddyl/core/actions/lqr.hxx:41]: I didn't realize this before.

We should cast a raw pointer, instead. This avoids to create an owner.
It should be something like this:
```c++
 ​  Data* d = ​static_cast​<Data*>(data.​get​());
```

**Review by cmastalli — COMMENTED** [review]
Thanks @wxmerkt! I just left a tiny extra comment.

**Review by cmastalli — APPROVED** [review]

**Comment by wxmerkt** [comment]
Thank you for handling the open comments. Let's merge this one after #1025 is merged

**Comment by cmastalli** [comment]
I am afraid that the Python bindings in #1025 might take you considerably time.
Perhaps, you could cherry pick the commits needed to allow the CI passes.
Thus, we don't need to wait for it.

**Comment by cmastalli** [comment]
@wxmerkt  -- thank you again for these fixes!

**Pull request merged**
